### ISSUE TITLE
Opinion: Recommendation to fork ConanCenter is bad advice

### ISSUE THREAD

**Issue by CJCombrink** [conan-io/docs#3368]
Why do I think forking Conan Center for production environments are a bad idea?
TLDR; Maintaining a fork is a time multiplier, it stands in the way of change, it undermines FOSS and gives a false sense of repeatability and reproducibility. 

### Time multiplier
Recently I contributed to CCI to upgrade the hunspell library to conan 2 ([CCI#18829](https://github.com/conan-io/conan-center-index/pull/18829)). This was 100% for a personal/hobby project. Just keeping my fork of CCI in sync was overhead I was not always willing/wanting to spend (personal time is valuable time). Even if at the least it involved pushing the "rebase" button on the PR for auto rebase or doing in manually in my local clone.

In a production environment there is almost always no time for managing libraries, we have to get features to the clients (in my experience at least). 

Packages and recipes in CCI changes so often, and almost all packages depends on a dozen more. To be on top of all of these changes all of the time, to be able to at some point update from upstream will be a full time job. 

This advice scares me: "_You can check and audit those changes if you want to, analyzing the diffs_". Package X depends on Y which depends on Z. What do you do with the audited outcome? Only apply a subset, to realize that a required change in dependency XXY is actually needed, fully unrelated to what you are trying to do (think openssl and zlib).

### It stands in the way of change
We are in a situation (pre revisions, pre migrating conan 2) where we are not allowed to touch versions because we know what we have works and does not have the time to figure out where it will break if we now upgrade a dependency. The last time we tried it broke everything. Just uploading a boost recipe caused _everything_ to conflict due to new dependencies that boost requires. 

When you fork CCI, you force people to put a peg in the ground, and there are a lot of reasons not to try and move that peg. 

### It undermines FOSS

Do you depend on a package or recipe that requires patching... Just apply it to our fork...
The central nature of CCI will enable people to keep their changes for themselves. The argument here is then but it would be in your best interest to also feed those change upstream to CCI for the community or else updating from upstream will get exponentially more difficult. Well if you are in any case not going to update from upstream then this does not apply, and it gives more reason for why updating from upstream will not happen. 

The recommendation on the [Creating and hosting your own ConanCenter binaries](https://docs.conan.io/2/devops/conancenter/hosting_binaries.html#devops-conancenter-hosting-your-own-conancenter-fork) even encourages local changes in your fork. 

We should not be naïve about people contributing upstream for the greater good: see the fallout with Red Hat about freeloading, the situation with HasiCorp, etc. The list is long about how OSS does not really work in practice, especially when there are deadlines and customers.

_Case study:_ We (our company) recently required changes to a package recipe. The actual plan was to create our own package internally, fix the recipe and upload internally while we wait for CCI to upgrade the package. We then decided against this but make sure we get CCI to upgrade and contribute where needed (in the spirit of OSS). Luckily for us the work was already in progress and we could just review, provide feedback and herd it to completion. This was a win for us few (and the conan team) since the sentiment in the team was it will take too long, do it internal. It proved to us and the team that getting packages fixed or updated through the official CCI is not painful.

### False sense of "Repeatability and Reproducibility"

One thing I have learned is not to trust anything on the internet, but then a local clone makes sense... No, because you can't clone the internet. Its almost an all or nothing...

 How many packages are changed on a weekly basis because the origin of the source changes? A URL change, repo rename, Sourceforce down, DNS breaks... You can't clone the internet. Cloning CCI adds a very small layer of protection, but what if you need to rebuild a package because a compiler or option changed? Your local clone points to a remote source and the remote source is down. You are stuck, unprotected.

An argument: I have read somewhere that conan have a clone or something of SourceForce (might have read it incorrectly). That covers one base for them, but not us, the users. What if they are down, again same boat. What if Github goes down? Do we also clone GitHub, Bitbucket, GitLab, etc. All or nothing... 

Case Study: Sourceforge was down 7 September 2023.

I am aware of the WIP docs on [Backup sources devops guide](https://github.com/conan-io/docs/pull/3365). From a quick read it looks promising, but I think will even add more of a time multiplier. Now changing a package requires a merge, rebuild, re-upload binaries, re-upload sources... oh wait Sourceforge is down so I can't get the new sources so that I can back them up.

## The competition

That other package manager followed this approach from the start (local clone... boils down to the same thing). We used it at some company for a while until we needed to change one dependency. Then abandoned the package manager and reverted to what we had before it. At that time even building the package manager was done from the local clone and that did not work.

Case Study: My personal project uses this package manager. This project is a plugin for the Qt Creator IDE and I make releases for each release of the IDE. About 80% of commits to my repo is just trying to get the package manager to build my (single) dependency. It is always an issue when I revisit the package on my local machine and need to rebuild. I am in a situation where I am not allowing myself to touch the clone that contains a working binary. That clone on my PC is more precious compared to any other info on that PC and I always back that up first when I do backups. I have personally gone through the effort to get my dependency conan 2 ready in CCI so that I can move to conan for my next release. I want to shout this: It is one dependency... for one project... for one developer... how can this be so difficult to reproduce.

_For some comic relief consider [this](https://github.com/CJCombrink/SpellChecker-Plugin/commit/0cae7bc78255d204e6bbff040554322597e4b950) beautiful commit just to get a release out since I could not get GitHub to build my dependency on Windows. At least I still had a binary on my local system that I could use as is (I could also net get it rebuild locally) to build the release package for Windows_ 
_The error: a good ol' mirror rename/move that were pointed to by the package manager:_
```
 error: Failed to download from mirror set
 error: https://repo.msys2.org/msys/x86_64/grep-3.0-2-x86_64.pkg.tar.xz: failed: status code 404
 error: https://www2.futureware.at/~nickoe/msys2-mirror/msys/x86_64/grep-3.0-2-x86_64.pkg.tar.xz: failed: status code 404
```

# Summary 

1. Cloning CCI should be discouraged for all workflows. 
2. Hosting your own binaries is a must.
    - Don't get me wrong: I am 100% against the local clone, but 110% for hosting at least the binaries that you depend on locally. 
    - If nothing changes and you can reproduce a known to be working build.
3. CCI should be the only master
4. Encourage contributing to CCI where changes are needed
    - JFrog/Conan is spending the effort and time to ensure that changes are reviewed and working.
    - They are picking up the pieces when stuff break (or stuff break each other)
    - They are working into backing up sources: pool resources there to make sure they get it right (or for them to get the resources to clone the internet)
    - If you have to audit changes: Do it there - review PRs, watch packages (maybe JFrog can make watching specific packages easier)
5. Changing packages will always be a pain in production (be it upgrade, options, fixes)
    - Don't give people more reason to work behind closed doors 
    - Don't make it more painful
    - Don't give people more excuses why it is too painful (or costly) to upgrade
 6. Be better and easier than the competition 
     - This comment on slack stands out for me: "Hi all, with X I could ismply clone the X repo and start using it""how is the workflow with conan? do I pip it?"
         - Lets make it easier, even for production

## Trivia

I have been developing C++ as my primary language for almost 15 years now. 15 years ago we had to write everything ourselves, even a package manager for internal libraries shared between projects. 

The other package manager made it better, then worse.

Two years ago I started at a company that was already invested in conan. We grew from about 5 or 6 internal libraries to almost 15 repos using conan. 

Since using conan I am convinced that conan is the answer to 15 yeas of struggle.

**Comment by samuel-emrys** [comment]
I don't want to take away from the point you've made, but I think it would be useful more widely to comment on alternative workflows.

One alternative to forking the conan-center-index repo is to just keep `conancenter` in your remotes on the build agent, and `conan install --requires mylib/x.y.z -o mylib/*:some_option=abc` for the set of configurations that you require. This way you can use the remote to fetch the recipe and build a binary for your custom configuration. You can then upload these binaries to your own conan local repository (`conancenter-builds`).

From a user perspective, if you add `conancenter` and `conancenter-builds` to an artifactory virtual repository `my-conancenter`, then when users search they will discover both the binaries from `conancenter` and `conancenter-builds` if `my-conancenter` is in their remotes list. Indeed, you can even do away with `conancenter` entirely by forcing builds/re-uploading binaries to `conancenter-builds`.

As far as modifying recipes go, you could host your own index internally. This doesn't solve all of the problems you present above, but it avoids the fork and provides a staging ground for future re-integration. This index would only be recipes of your own creation or the modifications of recipes that exist elsewhere. Configure a CI pipeline that builds all of the recipes int he configurations you need in a similar way to what conan-center-index do. When you mature the recipes here, they can then be moved back into or added to conan-center-index. Once they are added back to conan-center-index, you can deprecate your internal recipe as the maintenance burden has been shifted back to CCI.

Of course, as you've discovered, you can always engage to uplift recipes directly with conan-center-index. The above is more appropriate for recipes that are not suitable for CCI (i.e., closed source, or rely on custom generators), but it can be bent to that use case as well if you need a more responsive internal playground.

**Comment by rconde01** [comment]
A lot of interesting thoughts that I haven't fully digested...but a couple of things:
- There are some [issues](https://github.com/conan-io/conan/issues/12687) with using conan center as a virtual repository
- Surely not a solution for everyone, but my current sledgehammer approach to sources is to rebuild all packages for public releases and then zip/archive the entire conan cache.

**Comment by jcar87** [comment]
Hi @CJCombrink, thank you so much for raising this issue, and for reading the documentation and taking the time to write this opinion. We appreciate it has taken time to write and prepare.

I think that based on your feedback we can revisit and reframe some of the wording in our recently published guidelines: https://docs.conan.io/2/devops/using_conancenter.html.

Our intention is not to require nor advise all users of Conan Center to operate from a local clone or fork that is isolated from the mainline. 
Our guidelines have been written based on the feedback we have received over the years from our enterprise (hence "production") customers. The central idea is that the recipes in Conan Center remote are updated over time. For some consumers, it changes too quickly (new versions are made available, internal dependencies are updated, etc), while for others it does not change quickly enough. This can impact medium-to-large sized teams, where any disruption is very costly. Thus, our advice in those cases is to have better control that is local to the team in question. Not all companies have the same requirements, so the advice may not apply to all teams. For a lot of users (and smaller teams), operating directly with Conan Center and with lockfiles may be perfectly sufficient, and the guidelines are intended for them to take appropriate action based on their business needs. 

Regarding the contributions to FOSS, our experience with our existing users is that that they will happily both contribute Conan Center where relevant (it reduces the need to maintain deviations), and also host a copy with their changes locally, in case the changes are more urgent than the typical turnaround for a Conan Center PR. We are also aware that a lot of teams have fixes/patches/workarounds that are only relevant to them, and wouldn't be accepted in Conan Center or in the upstream libraries. From time to time we do reject contributions that are more "this suits the needs of my team", but are not suitable for the wider user base. A Conan Center PR also requires validation across a comprehensive platform matrix, and not all contributors have access to all these platforms and compiler versions, or the resources to dedicate. But in our experience, for patches that _are_ relevant bugfixes or features, users are happy to contribute, and use their locally hosted copy while the contribution goes through the regular process.

Regarding any issues that you've had in the past when updating or syncing up changes from Conan Center, we will happily look at specific issues reported, as we want to ensure that we are providing the right experience. While version conflicts can be frustrating, they are a valuable tool - failing early with a conflict that can be more or less easily resolved, tends to be a lot less work than experiencing compiler or linker errors down the line that are even more difficult to troubleshoot. But we are working towards ensuring that the default experience for Conan Center users is free of conflicts.
